### PR TITLE
Ninja build

### PR DESCRIPTION
**Pull request by em-eight**
Use ninja for C and ASM compilation

### CODE REVIEW

**Comment by terorie** [comment]
Nice work, finally.

**Comment by SeekyCt** [comment]
MWCC .d files don't work on linux, but dkp's cpp is used instead and should be able to do everything that they do already afaik

File: diff_settings.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def apply(config: dict, args):
         config["myimg"] = "artifacts/target/pal/main.dol"
         config["baseimg"] = "artifacts/orig/pal/main.dol"
     config["symbols_yml"] = "pack/symbols.yml"
-    config["make_command"] = [executable, "build.py", "--diff_py"]
+    config["make_command"] = "ninja"
```
Comment (SeekyCt) [inline diff_settings.py:20]: I think this needs to be `["ninja"]` btw

**Comment by em-eight** [comment]
All the suggestions are reasonable. One thing to note is that the remaining performance gain is by by quite a margin assembly regeneration, but we need to figure out and implement our own depfile. After that, ninja integration is pretty straightforward.

**Comment by SeekyCt** [comment]
Does it need anything more complicated than just depending on the slices and symbols files?

**Comment by em-eight** [comment]
The best thing would be to actually read them and generate only what's neccessary, it would save a lot of time.

**Comment by em-eight** [comment]
Btw, give me an hour or two before merging to add some of what terorie mentioned

**Comment by SeekyCt** [comment]
> The best thing would be to actually read them and generate only what's neccessary, it would save a lot of time.

I think that's something you'd have to do at the tool level rather than the ninja level though?

**Comment by em-eight** [comment]
> > The best thing would be to actually read them and generate only what's neccessary, it would save a lot of time.
> 
> I think that's something you'd have to do at the tool level rather than the ninja level though?

Yes, similar to how we use dkp gcc as a script to generate a depfile for C

**Comment by em-eight** [comment]
Note that the tools only generate the depfile, not the asm itself

**Comment by terorie** [comment]
> > The best thing would be to actually read them and generate only what's neccessary, it would save a lot of time.
> 
> I think that's something you'd have to do at the tool level rather than the ninja level though?

If we want to do it in Ninja, it requires one level of indirection. 
We could model a rule that generates a intermediate-step build.ninja from slices and symbols.txt.

For slices, this is easy – if the range changes, the resulting asm file name will also change and be rebuilt.

Reacting to symbol changes is a bit more difficult. symbols.txt is spans lots of asm files. So if the asm gen rule directly depends on it directly, it would regen everything if the modification date changes (e.g. a single line is added).
One workaround is to introduce a tool that splits symbols.txt into a file per object, which is cheap enough.
Then, let the object targets depend on their respective symbols split file.

Unlike Make, Ninja doesn't natively support recursion, so it would be hacky or require a wrapper script anyways.
It's probably not worth the effort.

**Comment by em-eight** [comment]
Probably easier to do at the tool level and reconfigure ninja yeah

**Comment by em-eight** [comment]
@SeekyCt Aight, PR is ready, I need some help with the windows CI though...

**Comment by SeekyCt** [comment]
It doesn't seem to like `-Isource/**`

**Comment by em-eight** [comment]
Perhaps there is some way to glob on windows? Maybe with a userspace utility?

**Comment by em-eight** [comment]
Otherwise this can be done in python https://stackoverflow.com/questions/2186525/how-to-use-glob-to-find-files-recursively

**Comment by SeekyCt** [comment]
Is there anything wrong with just doing source and source/platform like the CW args do?

**Comment by em-eight** [comment]
You're right

**Comment by em-eight** [comment]
`cc1.exe: error: bad value ('750') for '-mtune=' switch`
what

**Comment by em-eight** [comment]
@terorie @riidefi Just in case you didn't notice discord, the PR is ready for merge

**Comment by riidefi** [comment]
LGTM. Nice work!

**Pull request merged**